### PR TITLE
Use pytorch for 3D+ data

### PR DESCRIPTION
**Pull request by soupault**
Proof of concept.

- Switched to pytorch for 3D+ in Flip, Crop and Pad
- Flip now accepts multiple axes, but not axis=-1. Also, the axis order has been switched from opencv to numpy
- Minor cleanup to the tests

### CODE REVIEW

File: solt/core/_base_transforms.py [hunk 1]
```diff
@@ -99,9 +100,15 @@ def apply(self, data: DataContainer):
         for i, (item, t, item_settings) in enumerate(data):
             if i in self.data_indices:
                 if t == "I":  # Image
-                    tmp_item = self._apply_img(item, item_settings)
+                    if item.ndim <= 3 and isinstance(item, np.ndarray):
+                        tmp_item = self._apply_img(item, item_settings)
+                    else:
+                        tmp_item = self._apply_img_pt(item, item_settings)
```
Comment (lext) [inline solt/core/_base_transforms.py:106]: Let's not make this backend specific and call this `apply_img_nd`.

File: solt/core/_base_transforms.py [hunk 2]
```diff
@@ -99,9 +100,15 @@ def apply(self, data: DataContainer):
...
                 elif t == "M":  # Mask
-                    tmp_item = self._apply_mask(item, item_settings)
+                    if item.ndim <= 3 and isinstance(item, np.ndarray):
+                        tmp_item = self._apply_mask(item, item_settings)
+                    else:
+                        tmp_item = self._apply_mask_pt(item, item_settings)
```
Comment (lext) [inline solt/core/_base_transforms.py:111]: Let's not make this backend specific and call this `apply_mask_nd`.

File: solt/core/_base_transforms.py [hunk 3]
```diff
@@ -180,8 +187,8 @@ def __call__(
         return res
 
     @abstractmethod
-    def _apply_img(self, img: np.ndarray, settings: dict):
-        """Abstract method, which determines the transform's behaviour when it is applied to images HxWxC.
+    def _apply_img(self, img: np.ndarray, settings: dict) -> np.ndarray:
+        """Determines the transform applied to 2D mask (ndarray, last dim - channels).
```
Comment (lext) [inline solt/core/_base_transforms.py:191]: Typo in the doc. Not 2D mask, 2D image.

File: solt/core/_base_transforms.py [hunk 4]
```diff
@@ -210,6 +217,39 @@ def _apply_mask(self, mask: np.ndarray, settings: dict):
...
+        img : torch.Tensor
+            Image to be augmented
+
+        Returns
+        -------
+        out : torch.Tensor
+
+        """
+        msg = f"{self.__name__} does not currently supports torch.Tensors and 3D+ arrays"
+        raise NotImplementedError(msg)
```
Comment (lext) [inline solt/core/_base_transforms.py:234]: This method should be abstract, please remove `raise NotImplementedError(msg)`

File: solt/core/_data.py [hunk 5]
```diff
@@ -106,7 +106,7 @@ def validate(self):
                 if len(frame_prev) != len(frame_curr):
                     msg = f"Inconsistent frame dimensionality: " f"{len(frame_prev)}, {len(frame_curr)}"
                     raise ValueError(msg)
-                elif frame_prev != frame_curr:
+                elif frame_prev.tolist() != frame_curr.tolist():
```
Comment (lext) [inline solt/core/_data.py:109]: Shouldn't `np.array_equal` be better?

File: solt/transforms/_transforms.py [hunk 6]
```diff
@@ -589,6 +583,31 @@ def _apply_img(self, img: np.ndarray, settings: dict):
     def _apply_mask(self, mask: np.ndarray, settings: dict):
         return self._apply_img_or_mask(mask, settings)
 
+    def _apply_img_or_mask_pt(self, img_mask: torch.Tensor, settings: dict):
```
Comment (lext) [inline solt/transforms/_transforms.py:586]: I feel like the implementation of this function is not much different for 2D or ND. Should we call `_apply_img_or_mask_nd` for both opencv and torch backends, a and then via simple `if` detect which backend we should use?

File: solt/transforms/_transforms.py [hunk 7]
```diff
@@ -745,10 +764,16 @@ def __crop_img_or_mask(self, img_mask):
...
+        return self._apply_img_or_mask(img)
 
     def _apply_mask(self, mask: np.ndarray, settings: dict):
-        return self.__crop_img_or_mask(mask)
+        return self._apply_img_or_mask(mask)
+
+    def _apply_img_pt(self, img: torch.Tensor, settings: dict):
+        return self._apply_img_or_mask(img)
+
+    def _apply_mask_pt(self, mask: torch.Tensor, settings: dict):
```
Comment (lext) [inline solt/transforms/_transforms.py:775]: Similar here. Let's have `apply_img_or_mask_nd`.

**Comment by soupault** [comment]
Open points: cv2 2D, pt 3D, backend, device, gamma 3D, normalization

**Pull request closed without merging**